lib/rspec/rails/matchers/be_valid.rb

```diff
@@ -15,7 +15,8 @@ def matches?(actual)
         def failure_message
           message = "expected #{actual.inspect} to be valid"
 
-          if actual.respond_to?(:errors)
+          if actual.respond_to?(:errors) &&
+              (!actual.methods.include?(:errors) || actual.method(:errors).arity < 1)
```

Comment:
You don't need this duplicate methods check, if `if actual.respond_to?(:errors) && actual.method(:errors).arity < 1` will suffice.

Reply:
I added the `methods.include?` check because, given how many ways there are to define functions in ruby, I figured there was probably some way or another to have an object not work with the original check.

When I actually tried to construct such an object, I found that it _is_ possible, but it's pretty edge-casey:

```
class A
  def respond_to_missing?(name, _)
    return true if name.to_sym == :foo
    super
  end
end
a = A.new
a.respond_to?(:foo) # true
a.method(:foo).arity # -1
a.foo # NoMethodError: undefined method `foo' for #<A:0x00007fd60c168430>
a.methods.include?(:foo) # false
```

So with a class defined like that, we could still run into errors in this code without the `.include?` check.  Although if you think that's too much of an edge-case to bother accounting for, I'm happy to remove it.

Reply:
We don't support classes that don't behave according to ruby semantics, if someone uses the trick you mention it won't work with the matcher and thats ok, it'll just return false. If you add the correct `method_missing` definition to go with that snippet the method should be returned.